### PR TITLE
8303355: The Depend plugin does fully recompile when primitive type changes

### PR DESCRIPTION
**Pull request by lahodaj**
The OpenJDK build is using a Plugin called Depend to avoid building Java code unnecessarily. It has two parts, one is checking module APIs (and forces rebuild of dependent modules if a dependency changes), and second takes modified files in a module, and attempts to detect whether it is enough to recompile these files, or if the whole module must be rebuilt.

There's a bug in the second part, it does not track changes in primitive types, so e.g. a change of a type of a public field form `int` to `long` does not cause recompile.

This patch fixes that by visiting the primitive types, and including them in the hash computed for the given file.

There's also another problem - if a module is being recompiled from scratch due to a change in a file hash, the new/updated file hashes are written immediately. And if the compilation consequently fails, the state is not compared to the original hash, but to the hash from the broken compilation, which may lead to missing compilation of some files.

This patch moves the code to write the new hashes to the end of compilation, and only do that when there were no compile-time errors during the compilation.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303355](https://bugs.openjdk.org/browse/JDK-8303355): The Depend plugin does fully recompile when primitive type changes

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [63cd6884](https://git.openjdk.org/jdk/pull/12801/files/63cd688419b15f234bb49b9301fe9347ae854821)
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12801/head:pull/12801` \
`$ git checkout pull/12801`

Update a local copy of the PR: \
`$ git checkout pull/12801` \
`$ git pull https://git.openjdk.org/jdk pull/12801/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12801`

View PR using the GUI difftool: \
`$ git pr show -t 12801`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12801.diff">https://git.openjdk.org/jdk/pull/12801.diff</a>

</details>

### CODE REVIEW

File: make/jdk/src/classes/build/tools/depend/Depend.java [hunk 1]
```diff
@@ -878,24 +880,30 @@ public Void visitModifiers(ModifiersTree node, Void p) {
             return super.visitModifiers(node, p);
         }
 
+        @Override
+        public Void visitPrimitiveType(PrimitiveTypeTree node, Void p) {
+            update(node.getPrimitiveTypeKind().name());
```
Comment (vicente-romero-oracle) [inline make/jdk/src/classes/build/tools/depend/Depend.java:885]: shouldn't `DependTest` include a test to stress this new code?

**Review by erikj79 — APPROVED** [review]

File: make/jdk/src/classes/build/tools/depend/Depend.java (hunk 1 above)
Comment (lahodaj) [reply make/jdk/src/classes/build/tools/depend/Depend.java:885]: Test updated:
https://github.com/openjdk/jdk/pull/12801/commits/de84439d64f6882d64f657fee0f6c55b2f717a9c

Thanks!

**Review by vicente-romero-oracle — APPROVED** [review]
looks good, thanks

**Pull request closed without merging**